### PR TITLE
#68-adding-amazon-to-extract

### PR DESCRIPTION
**Pull request by jk-george**
Closes: [Issue 68](https://github.com/users/gem09lo/projects/1/views/1?pane=issue&itemId=89987106&issue=gem09lo%7CPriceSlashTrack%7C68).

## Outcome:
- The extract.py script has a new function called scrape_pricing_process
- This function decides which scrape function to use in order to scrape the given URL (if it's Amazon/Steam it will use a respective function to scrape, if not it logs an error)
- Added a new function to scrape amazon website product pages and find the correct prices.
- Allowed schema.sql to have more characters in the product_name

### CODE REVIEW

File: pipeline/extract.py [hunk 1]
```diff
@@ -85,20 +87,72 @@ def get_website_from_url(url: str) -> str:
...
+
+def scrape_from_amazon_html(html_content, url, product_id):
     """ Scrapes from html to get a dictionary with the:
     - Product_ID 
     - product_name
     - original_price
     - discount_price
     - website 
-    """
+    for Amazon Product Pages."""
```
Comment (gem09lo) [inline pipeline/extract.py:113]: Are you able to shorten this docstring? e.g. "Scrapes product, price and website information from Amazon."

File: pipeline/extract.py [hunk 2]
```diff
@@ -85,20 +87,72 @@ def get_website_from_url(url: str) -> str:
...
+
+    if "https://www.amazon.co" in url:
+        return scrape_from_amazon_html(html_content, url, product_id)
+
+    logging.error(
+        "Cannot scrape that URL, since it's not an Amazon/Steam webpage.")
+    return
+
+
+def scrape_from_amazon_html(html_content, url, product_id):
```
Comment (gem09lo) [inline pipeline/extract.py:106]: can you add type hints

File: pipeline/extract.py [hunk 3]
```diff
@@ -85,20 +87,72 @@ def get_website_from_url(url: str) -> str:
...
+
+
+def scrape_from_steam_html(html_content: bytes, url: str, product_id: int) -> dict:
+    """ Scrapes from html to get a dictionary with the:
+    - Product_ID 
+    - product_name
+    - original_price
+    - discount_price
+    - website 
+    for STEAM Games/Products."""
```
Comment (gem09lo) [inline pipeline/extract.py:149]: Can this also be shortened? e.g. "Scrapes product, price and website information from Steam."

File: pipeline/extract.py [hunk 4]
```diff
@@ -85,20 +87,72 @@ def get_website_from_url(url: str) -> str:
...
+    for Amazon Product Pages."""
+    s = BeautifulSoup(html_content, 'html.parser')
+
+    results = s.find("div", id="corePriceDisplay_desktop_feature_div")
+
+    if not results:
+        logging.error("Can't scrape from Amazon URL")
+        return None
+
+    game_title_elem = s.find(id="productTitle")
```
Comment (gem09lo) [inline pipeline/extract.py:122]: Can this be renamed to something else - This variable name is specific to Steam products. Can rename it to product_name_element or something.

File: pipeline/extract.py [hunk 5]
```diff
@@ -85,20 +87,72 @@ def get_website_from_url(url: str) -> str:
...
+
+    if not game_title_elem:
```
Comment (gem09lo) [inline pipeline/extract.py:124]: I didn't pick it up last time but following the code style guide we should avoid abbreviations.

File: pipeline/test_extract.py [hunk 6]
```diff
@@ -1,6 +1,12 @@
 import pytest
```
Comment (gem09lo) [inline pipeline/test_extract.py:1]: Nice tests!

File: pipeline/test_extract.py [hunk 7]
```diff
@@ -1,6 +1,12 @@
...
 
-from extract import scrape_from_html, get_html_from_url, get_website_from_url
+from extract import scrape_from_steam_html, get_html_from_url, get_website_from_url, scrape_pricing_process
+
+
+def test_scrape_pricing_process_only_accepts_amazon_and_steam():
```
Comment (gem09lo) [inline pipeline/test_extract.py:6]: Good function names!

File: pipeline/test_extract.py [hunk 8]
```diff
@@ -1,6 +1,12 @@
...
+    """Tests that scrape pricing process only accepts amazon and steam"""
```
Comment (gem09lo) [inline pipeline/test_extract.py:7]: And descriptive docstrings

File: schema.sql [hunk 9]
```diff
@@ -21,7 +21,7 @@ CREATE TABLE website (
 
 CREATE TABLE product (
     product_id INT GENERATED ALWAYS AS IDENTITY,
-    product_name VARCHAR(100) NOT NULL,
+    product_name VARCHAR(250) NOT NULL,
```
Comment (gem09lo) [inline schema.sql:24]: Nice spot

File: test_database.sql [hunk 10]
```diff
@@ -32,7 +32,9 @@ VALUES
     ('Smartwatch', 'https://www.target.com/smartwatch', 5, 149.99),
     ('Slay the Spire', 'https://store.steampowered.com/app/646570/Slay_the_Spire/', 6, 19.99),
     ('Cyberpunk 2077','https://store.steampowered.com/app/1091500/Cyberpunk_2077/',6,49.99),
-    ('Stardew Valley','https://store.steampowered.com/app/413150/Stardew_Valley/',6,10.99);
+    ('Stardew Valley','https://store.steampowered.com/app/413150/Stardew_Valley/',6,10.99),('Amazon Basics 3-Button USB Wired Quiet Mouse – Standard, Black','https://www.amazon.co.uk/Amazon-Basics-3-Button-Wired-Quiet/dp/B08P6FXKP9/ref=sr_1_1_ffob_sspa?dib=eyJ2IjoiMSJ9.g3__HWg7j0aKkKDxSF-bzRPowngP20NVGO9Qu_tFElQOss-j4y20Zsap8Hmjp8txZBteeMiJZwTjdUKUi5qmLNcIf6CzaKEGVtjw99C1Z6TA_pVy-IFgYyd7ilktGnUuGfGEYhyip4LF4aiCwDqkQZcHPl7UJOz33jY0F73lAWqhAsC2DTZK2nCpWfuU4f6Bhb6qK--Q9Esz6xgQ6DUaIYHHp5LJDcUlhlecP1gCDSU.ZfZC0Snr23MWdCJfl7Y2IXcUHMblQiow8boVuUdIs7g&dib_tag=se&keywords=mouse&qid=1733492153&sr=8-1-spons&sp_csd=d2lkZ2V0TmFtZT1zcF9hdGY&psc=1',1,8.99)
+    ('Razer Viper V3 Pro','https://www.amazon.co.uk/Razer-Viper-Pro-Ultra-lightweight-HyperPolling/dp/B0CSPN2G5Z/',1,159.99);
+
```
Comment (gem09lo) [inline test_database.sql:37]: Good - more test data to work with. Im assuming this is products 12 and 13?

**Review by gem09lo — COMMENTED** [review]
Overall good code! Just a few minor changes for code quality. 
I'd want India to do a second review and approve it as I am not as familiar with beautiful soup as you guys are.

File: pipeline/extract.py [hunk 11]
```diff
@@ -85,20 +87,72 @@ def get_website_from_url(url: str) -> str:
     return website_url
 
 
-def scrape_from_html(html_content: bytes, url: str, product_id: int) -> dict:
+def scrape_pricing_process(html_content: bytes, url: str, product_id: int) -> dict:
```
Comment (indiah444) [inline pipeline/extract.py:90]: Great function, makes sure we're scraping the right links

File: pipeline/extract.py (hunk 5 above)
Comment (indiah444) [reply pipeline/extract.py:124]: Ah, I didn't realise that, I was the one who used abbreviations first, apologies!

**Review by indiah444 — CHANGES_REQUESTED** [review]
Great work so far, just a few small things should be changed for quality/consistency's sake (as mentioned in the comments before), but overall, the functionality of this is great.

File: pipeline/extract.py (hunk 1 above)
Comment (jk-george) [reply pipeline/extract.py:113]: Done.

**Review by indiah444 — APPROVED** [review]
This all looks great, thank you for making these changes!

File: .github/workflows/pytest.yml [hunk 12]
```diff
@@ -23,13 +23,10 @@ jobs:
...
-        run: |
-          if [ -f pytest.ini ] || [ -f pyproject.toml ] || [ -d tests ]; then
-            coverage run -m pytest -v -s || echo "Tests failed or skipped."
-          else
-            echo "No test configuration or tests folder found, skipping pytest."
-          fi
+      - name: Downloading pytest
+        run: pip3 install pytest
+      - name: Testing with pytest
+        run: pytest
```
Comment (indiah444) [inline .github/workflows/pytest.yml:29]: Thank you very much for fixing this, this is fantastic!

Comment (jk-george) [reply .github/workflows/pytest.yml:29]: :)

File: dashboard.py [hunk 13]
```diff
@@ -68,49 +68,94 @@ def get_website_from_url(url: str) -> str:
...
 
     results = s.find(id="game_area_purchase")
 
     if not results:
-        logging.error("Can't scrape that URL.")
+        logging.error("Can't scrape that Steam URL.")
         return None
 
-    original_price_elem = results.find(
+    original_price_element = results.find(
```
Comment (indiah444) [inline dashboard.py:128]: Apologies for using these abbreviations in the first place, thanks for fixing them! 😅

File: database/schema.sql [hunk 14]
```diff
@@ -22,7 +22,7 @@ CREATE TABLE website (
 
 CREATE TABLE product (
     product_id INT GENERATED ALWAYS AS IDENTITY,
-    product_name VARCHAR(100) NOT NULL,
+    product_name VARCHAR(250) NOT NULL,
```
Comment (indiah444) [inline database/schema.sql:25]: Great fix to increase the length of the product name

**Comment by jk-george** [comment]
## Bug fixes:
The following bugs were fixed within this branch:
- [Issue 102](https://github.com/users/gem09lo/projects/1?pane=issue&itemId=90398300&issue=gem09lo%7CPriceSlashTrack%7C102)
- [Issue 103](https://github.com/users/gem09lo/projects/1/views/1?pane=issue&itemId=90398613&issue=gem09lo%7CPriceSlashTrack%7C103)
- [Issue 104](https://github.com/users/gem09lo/projects/1/views/1?pane=issue&itemId=90399070&issue=gem09lo%7CPriceSlashTrack%7C104)

## Technical Issue Found:
- [Technical Issue 105](https://github.com/users/gem09lo/projects/1/views/1?pane=issue&itemId=90399350&issue=gem09lo%7CPriceSlashTrack%7C105)

### Pull request will still move forward and be merged, even with this technical issue.

**Pull request merged**